pkg/cmd/checkendpoints/controller/connection_checker.go

```diff
@@ -122,18 +128,95 @@ func (c *connectionChecker) checkEndpoint(ctx context.Context, check *operatorco
 }
 
 // getTCPConnectLatency connects to a tcp endpoint and collects latency info
-func getTCPConnectLatency(ctx context.Context, address string) (*trace.LatencyInfo, error) {
+func (c *connectionChecker) getTCPConnectLatency(ctx context.Context, address string) (*trace.LatencyInfo, error) {
 	klog.V(4).Infof("Check BEGIN: %v", address)
 	defer klog.V(4).Infof("Check END  : %v", address)
 	ctx, latencyInfo := trace.WithLatencyInfoCapture(ctx)
-	dialer := &net.Dialer{Timeout: 10 * time.Second}
-	conn, err := dialer.DialContext(ctx, "tcp", address)
+	dialer := &net.Dialer{
+		Timeout: 10 * time.Second,
+	}
+	conn, err := DialWithDialerContext(ctx, dialer, "tcp", address, &tls.Config{
+		Certificates:       c.clientCerts,
+		InsecureSkipVerify: true,
+	})
 	if err == nil {
 		conn.Close()
 	}
+	// ignore tls connection error. most likely non-tls connection, plus we're not really testing tls
+	if _, ok := err.(tls.RecordHeaderError); ok {
+		klog.V(4).Infof("%s: tls error ignored: %v", address, err)
+		return latencyInfo, nil
+	}
 	return latencyInfo, err
 }
 
+// 	DialWithDialerContext is a copy of tls.DialWithDialer that uses the provided context.
+func DialWithDialerContext(ctx context.Context, dialer *net.Dialer, network, addr string, config *tls.Config) (*tls.Conn, error) {
+	// We want the Timeout and Deadline values from dialer to cover the
+	// whole process: TCP connection and TLS handshake. This means that we
+	// also need to start our own timers now.
+	timeout := dialer.Timeout
+
+	if !dialer.Deadline.IsZero() {
+		deadlineTimeout := time.Until(dialer.Deadline)
+		if timeout == 0 || deadlineTimeout < timeout {
+			timeout = deadlineTimeout
+		}
+	}
+
+	var errChannel chan error
+
+	if timeout != 0 {
```

Comment:
why should our code allow no timeout. I prefer forcing a default timeout if none is specified.

Reply:
> why should our code allow no timeout. I prefer forcing a default timeout if none is specified.

we should prevent this. Always default the timeout to something so we don't need this if block.